s3/pkg/service/object.go

```diff
@@ -93,165 +92,131 @@ func (dr *StreamReader) Read(p []byte) (n int, err error) {
-	oldObj, err := s.MetaStorage.GetObject(ctx, obj.BucketName, obj.ObjectKey, "", true)
-	if err != nil && oldObj != nil {
-		log.Info("got the object with same name(%s, %s, %s)", oldObj.BucketName, oldObj.ObjectKey, oldObj.ObjectId)
-		obj.StorageMeta = oldObj.StorageMeta
-		obj.ObjectId = oldObj.ObjectId
-	}
 
-	bucket, err := s.MetaStorage.GetBucket(ctx, obj.BucketName, true)
+	log.Infof("*********bucket:%s,key:%s,size:%d\n", req.BucketName, req.ObjectKey, req.Size)
+	bucket, err := s.MetaStorage.GetBucket(ctx, req.BucketName, true)
 	if err != nil {
 		log.Errorln("get bucket failed with err:", err)
 		return err
 	}
-	if bucket == nil {
-		log.Infoln("bucket is nil")
-	}
 
-	switch bucket.Acl.CannedAcl {
-	case "public-read-write":
-		break
-	default:
-		if bucket.TenantId != obj.TenantId {
-			return s3error.ErrBucketAccessForbidden
+	if !isAdmin {
+		switch bucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if bucket.TenantId != tenantId {
+				err = ErrBucketAccessForbidden
+				return err
+			}
 		}
 	}
 
+	err = s.removeObject(ctx, bucket, req.ObjectKey)
+	if err != nil {
+		log.Errorln("failed to delete old object. err:", err)
+		return err
+	}
+
 	data := &StreamReader{in: in}
 	var limitedDataReader io.Reader
-	if obj.Size > 0 { // request.ContentLength is -1 if length is unknown
-		limitedDataReader = io.LimitReader(data, obj.Size)
+	if req.Size > 0 { // request.ContentLength is -1 if length is unknown
+		limitedDataReader = io.LimitReader(data, req.Size)
 	} else {
 		limitedDataReader = data
 	}
 
 	backendName := bucket.DefaultLocation
-	if obj.Location != "" {
-		backendName = obj.Location
+	if req.Location != "" {
+		backendName = req.Location
 	}
 	backend, err := utils.GetBackend(ctx, s.backendClient, backendName)
 	if err != nil {
 		log.Errorln("failed to get backend client with err:", err)
 		return err
 	}
 
-	log.Infoln("bucket location:", obj.Location, " backendtype:", backend.Type,
-		" endpoint:", backend.Endpoint)
-	bodyMd5 := md["md5Sum"]
+	log.Infoln("bucket location:", req.Location, " backendtype:", backend.Type, " endpoint:", backend.Endpoint)
+	bodyMd5 := req.Attrs["md5Sum"]
 	ctx = context.Background()
-	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_SIZE, obj.Size)
+	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_SIZE, req.Size)
 	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_MD5, bodyMd5)
 	sd, err := driver.CreateStorageDriver(backend.Type, backend)
 	if err != nil {
 		log.Errorln("failed to create storage. err:", err)
-		return nil
+		return err
 	}
-	res, err := sd.Put(ctx, limitedDataReader, obj)
+	res, err := sd.Put(ctx, limitedDataReader, &pb.Object{BucketName:req.BucketName, ObjectKey:req.ObjectKey})
 	if err != nil {
 		log.Errorln("failed to put data. err:", err)
 		return err
 	}
-	oid := res.ObjectId
-	bytesWritten := res.Written
+	var delObj *pb.DeleteObjectInput
+	defer s.removeObjectFromBackend(ctx, sd, delObj)
```

Comment:
Is it better to call the function directly instead of using defer?

Reply:
@sfzeng , if there are more than one error where we need delete written object, it is better to use defer?